drivers: uart: nrfx_uarte: Major rework of the legacy UART shim to allow using it for nrf54x

nordic-krch opened the pull request:
`uart_nrfx_uarte2` shim which is using `nrfx_uarte` driver has proven to be not a good approach because the split between nrfx and the shim (and nrfx driver supporting additional APIs beyond zephyr one) increases memory footprint and decreases performance and makes it hard to maintain the shim. With significantly increased size it was not feasible to use that shim on cores with small memory (like `cpuppr` on nrf54h20).

Legacy shim RX asynchronous path had to be reworked because we cannot rely on RXDRDY events when counting bytes because EasyDMA transfers bytes in words so triggering RXDRDY event does not necessary mean that byte is available in RAM.

Rework driver to support new way of asynchronous RX handling. Previously RX was handled in two modes: using RXDRDY interrupt for byte counting or TIMER + PPI. Both modes had flaws. RXDRDY interrupt mode could miscalculated amount of received bytes when interrupt was not  handled on time. Data was not lost but was not reported on time and that could lead to issues. PPI+TIMER mode requires additional resources  thus it was not the default mode. Often user was not aware of that option and was experiencing driver RX faults.

New RX mode is switching buffers when there is new data (RXDRDY event not set for given amount of time). It does not require additional resources to get precise byte counting. Additionally, this is in line with new UARTE feature (RX frame timeout) which is present in nRF54X devices. The behavior of the driver is the same for legacy devices and new one. For legacy devices k_timer periodic interrupts are used to check if there are any new bytes and it is not needed when RX frame timeout is present.

Improved RX mode is enabled by default (CONFIG_UART_NRFX_UARTE_ENHANCED_RX=y) but legacy modes are still available though not recommended to be used.

Note that new RX mode behaves a bit different because timeout always triggers switch of buffers which means that there will be no UART_RX_RDY events with non-zero offset. It also means that every UART_RX_RDY will be followed by UART_RX_BUF_RELEASED.

Driver structures are reworked and asynchronous data is split into two structures: tx and rx data.

Additionally, support for DMM is added to asynchronous RX. For TX a fixed size, statically allocated buffer is used. Same goes for single byte buffers used for polling and interrupt driven mode (and RX FIFO flush buffer).

After rework, driver is recommended to be used for all platforms as it performs much better (2 times lower CPU load during same data) and takes much less code than the second UART shim available for Nordic devices.

PR contains tests added in #71886 #73638 and size optiomization from #74849.

nordic-krch committed:
tests: drivers: uart: Add test for async API using two instance

Add test which is using two independent UART devices. Validate behavior of
asynchronous API.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
boards: nordic: nrf54h20dk: Disable asserts on cpuppr for tests

cpuppr is a small core with less than 64k of memory. In order to
fit tests into that memory by default disable asserts (which saves
few kB).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Add nrf54h20 cpuppr

Add configuration for nrf54h20dk/nrf54h20/cpuppr.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: mix_fifo_poll: Add nrf54h20 cpuppr

Add configuration for nrf54h20dk/nrf54h20/cpuppr.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Rework driver to support new features

Rework driver to support new way of asynchronous RX handling.
Previously RX was handled in two modes: using RXDRDY interrupt for byte
counting or TIMER + PPI. Both modes had flaws. RXDRDY interrupt mode
could miscalculated amount of received bytes when interrupt was not
handled on time. Data was not lost but was not reported on time that
could lead to issues. PPI+TIMER mode requires additional resources
thus it was not the default mode. Often user was not aware of that
option and was expiriencing driver RX faults.

New RX mode is switching buffers when there is new data (RXDRDY event
not set for given amount of time). It does not require additional
resources to get precise byte counting. Additionally, this is in line
with new UARTE feature (RX frame timeout) which is present in nRF54X
devices. The behavior of the driver is the same for legacy devices
and new one. For legacy devices k_timer periodic interrupts are used
to check if there are any new bytes and it is not needed when RX frame
timeout is present.

Improved RX mode is enabled by default
(CONFIG_UART_NRFX_UARTE_ENHANCED_RX=y) but legacy modes are still
available though not recommended to be used.

Note that new RX mode behaves a bit different because timeout always
triggers switch of buffers which means that there will be no
UART_RX_RDY events with non-zero offset. It also means that every
UART_RX_RDY will be followed by UART_RX_BUF_RELEASED.

After rework, driver is recommended to be used for all platforms as it
performs much better and takes much less code than the second UART shim
available for Nordic devices.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Improve RX FIFO flush workaround

Add configurable magic byte instead of fixed 0xAA. In some cases, it
is known that certain bytes are less likely in the transmission and
picking specific magic byte may reduce probability of failure of
detection of the correct amount of flushed data.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Add uart120 instance to nrf54h20dk

Add second instance to be tested on nrf54h20dk. uart120 is a fast UARTE
which works on fixed pin locations. It is not available for cpuppr core.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Rework for multi instance

Rework the test to be able to run on multiple instances.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Add missing static keyword

Multiple variable in the test were missing static keyword.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: mix_fifo_poll: Rework to support multi-instance

Rework test to allow testing multiple UART instances in a single
run.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: mix_fifo_poll: Add UARTE120 to nrf54h20dk test

Add configuration for UARTE120 fast instance. Use multi-instance capability
of the test to run same test on UARTE120 and slow UARTE peripheral.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_dual: Rework for multi-instance

Rework test to be able to support running on multiple
set of instance(s).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_dual: Add testing of UARTE120

Use multi-instance testing and add UARTE120 configuration to nrf54h20
overlays.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Fix nrf54h20dk cpurad

Fix target which was failing due to lack of memory for DMA
buffers.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add runtime PM

Add runtime PM to the driver. When asynchronous or interrupt
driven API is used, there are 3 independent paths for enabling
the device: RX, TX and TX poll_out. TX poll_out requires special
handling because number of poll_out calls does not need to much
number of TXSTOPPED interrupts. To handle that special flag is
added. For standard RX and TX is simpler.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add support for DMM and cache

Add support for DMM which manages cache and dedicated memory spaces.
Added support for data cache for buffers which are not DMM managed.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_dual: Add test cases with runtime PM

Add configurations in which runtime PM is enabled. There are 2
configurations:
- Test does not call PM API for getting and putting the device.
 In that configuration UART device shall be suspended whenever
 not active.
- Test is calling PM API which should keep device active as
 long as requested (e.g. during packet reception which consists
 of multiple chunks).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add workaround for BAUDRATE register

On uart120 BAUDRATE register is not retained when ENABLE=0 and
because of that BAUDRATE must be set after enabling. Add workaround
to the driver.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add lock to uart_rx_disable

Add lock to fix race when uart_rx_disable is interrupted by RXTO
event which lead to driver state corruption.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

bjarki-andreasen commented on dts/bindings/serial/nordic,nrf-uarte.yaml:
```diff
@@ -9,3 +9,7 @@ properties:
     type: boolean
     description: |
       UARTE has ENDTX_STOPTX HW short.
+  frame-timeout-supported:
```
please add a newline between properties :)

bjarki-andreasen reviewed (commented):
The additions look great, but I do believe this PR is too large to be properly reviewed by anyone without hours to set aside. Especially regarding the new dual and pm test suites, which really need their own PRs and discussions. For example, the pm test suite should likely also include other vendors UARTs right? and we should agree on how PM is expected to work.

Could you please create a PR just with the update to the uart driver, maybe even excluding the addition of PM to it? otherwise I really feel we won't be able to discuss these additions properly

nordic-krch commented:
@bjarki-andreasen yes, i know that it is huge. I tried to split changes in the driver into multiple commits. I can open PR with just driver changes and keep this PR with all things (as i use all those tests to validate that driver works in all scenarios). NCS is waiting for that changes so would be good to move forward with review in next few weeks.

nordic-krch committed:
drivers: serial: nrfx_uarte: Remove CONFIG_UART_n_GPIO_MANAGEMENT

This is a leftover from pre-pinctrl era and no longer makes sense.
Driver always manages gpio through pinctrl.

Support removed from uart and uarte shims.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@bjarki-andreasen I've opened #78887 which contains main change from that PR (without test changes and additional changes in the driver).

nordic-krch committed:
drivers: serial: nrfx: Use legacy shim by default

So far new platforms did not use legacy shim but it shall
be now a default uart shim for all platforms.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

kl-cruz commented on tests/drivers/uart/uart_mix_fifo_poll/prj.conf:
```diff
@@ -1,11 +1,17 @@
 
-# CONFIG_NO_OPTIMIZATIONS=y
+CONFIG_OUTPUT_DISASSEMBLY=y
+#CONFIG_NO_OPTIMIZATIONS=y
```
I think some of configs here can be removed. Additionally, do we need disassembly output here?

nordic-krch replied:
This PR becomes an superset of all changes (uart + tests). I will be opening clean PR with only driver changes (PM) since that is what we need now. Tests will come later with less pressure.
